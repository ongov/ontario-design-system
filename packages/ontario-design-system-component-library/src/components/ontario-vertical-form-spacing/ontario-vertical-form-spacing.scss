@use '@ongov/ontario-design-system-global-styles/dist/styles/scss/1-variables/spacing.variables' as spacing;

// Mixin for generating individual vertical form spacing classes
@mixin ontario-vertical-form-spacing($modifier, $spacing-variable) {
	.ontario-vertical-form-spacing--#{$modifier},
	.ontario-vertical-form-spacing--#{$modifier} form {
		::slotted(ontario-input:not(:last-child)),
		::slotted(ontario-checkboxes:not(:last-child)),
		::slotted(ontario-dropdown-list:not(:last-child)),
		::slotted(ontario-textarea:not(:last-child)),
		::slotted(ontario-radio-buttons:not(:last-child)),
		::slotted(ontario-button:not(:last-child)),
		::slotted(ontario-date-input:not(:last-child)) {
			margin-bottom: (#{$spacing-variable});
			display: block;
		}
	}
}

// Mixin for generating nested vertical form spacing classes
@mixin ontario-nested-vertical-form-spacing($parent-modifier, $parent-spacing, $child-spacings) {
	.ontario-vertical-form-spacing--#{$parent-modifier},
	.ontario-vertical-form-spacing--#{$parent-modifier} form {
		@each $child-spacing in $child-spacings {
			$child-modifier: nth($child-spacing, 1);
			$child-spacing-variable: nth($child-spacing, 2);

			.ontario-vertical-form-spacing--#{$child-modifier},
			.ontario-vertical-form-spacing--#{$child-modifier} form {
				::slotted(ontario-input:not(:last-child)),
				::slotted(ontario-checkboxes:not(:last-child)),
				::slotted(ontario-dropdown-list:not(:last-child)),
				::slotted(ontario-textarea:not(:last-child)),
				::slotted(ontario-radio-buttons:not(:last-child)),
				::slotted(ontario-button:not(:last-child)),
				::slotted(ontario-date-input:not(:last-child)) {
					margin-bottom: $child-spacing-variable;
					display: block;
				}

				// Add extra spacing for the last child
				::slotted(ontario-input:last-child),
				::slotted(ontario-checkboxes:last-child),
				::slotted(ontario-dropdown-list:last-child),
				::slotted(ontario-textarea:last-child),
				::slotted(ontario-radio-buttons:last-child),
				::slotted(ontario-button:last-child),
				::slotted(ontario-date-input:last-child) {
					margin-bottom: $child-spacing-variable + $parent-spacing;
					display: block;
				}
			}
		}
	}
}

// Vertical spacing definitions
$vertical-spacing-definitions: (16, spacing.$spacing-4), (40, spacing.$spacing-7);

// Generate styles for each vertical spacing
@each $spacing-definition in $vertical-spacing-definitions {
	$spacing-modifier: nth($spacing-definition, 1);
	$spacing-variable: nth($spacing-definition, 2);

	// Generate individual vertical form spacing classes
	@include ontario-vertical-form-spacing($spacing-modifier, $spacing-variable);

	// Generate nested vertical form spacing classes
	$nested-spacings: ();
	@each $nested-spacing in $vertical-spacing-definitions {
		@if $spacing-modifier != nth($nested-spacing, 1) {
			$nested-spacings: append($nested-spacings, $nested-spacing);
		}
	}

	@include ontario-nested-vertical-form-spacing($spacing-modifier, $spacing-variable, $nested-spacings);
}
