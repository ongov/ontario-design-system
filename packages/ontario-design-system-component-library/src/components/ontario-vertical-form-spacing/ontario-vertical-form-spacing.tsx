import { Component, Element, h, Prop } from '@stencil/core';

@Component({
	tag: 'ontario-vertical-form-spacing',
	styleUrl: 'ontario-vertical-form-spacing.scss',
	shadow: true,
})
export class OntarioVerticalFormSpacing {
	@Element() element: HTMLElement;

	/**
	 * Defines the spacing between slotted form elements.
	 * If no spacing prop is provided, it will default to 'default'.
	 */
	@Prop() spacing: 'default' | 'condensed' = 'default';

	/**
	 * Returns the pixel value for the selected spacing option.
	 */
	private getSpacingValue() {
		const spacingOptions = {
			default: '40',
			condensed: '16',
		};

		return spacingOptions[this.spacing] || spacingOptions.default;
	}

	render() {
		return (
			<div class={`ontario-vertical-form-spacing--${this.getSpacingValue()}`}>
				<slot />
			</div>
		);
	}
}
