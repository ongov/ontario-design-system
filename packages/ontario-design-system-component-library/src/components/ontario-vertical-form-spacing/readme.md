# ontario-vertical-form-spacing

<!-- Auto Generated Below -->

## Properties

| Property  | Attribute | Description                                                                                                      | Type                       | Default     |
| --------- | --------- | ---------------------------------------------------------------------------------------------------------------- | -------------------------- | ----------- |
| `spacing` | `spacing` | Defines the spacing between slotted form elements. If no spacing prop is provided, it will default to 'default'. | `"condensed" \| "default"` | `'default'` |

---

_Built with [StencilJS](https://stenciljs.com/)_
