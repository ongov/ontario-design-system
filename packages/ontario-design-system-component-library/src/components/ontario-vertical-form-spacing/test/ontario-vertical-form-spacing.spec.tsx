import { newSpecPage } from '@stencil/core/testing';
import { OntarioVerticalFormSpacing } from '../ontario-vertical-form-spacing';

describe('ontario-vertical-form-spacing', () => {
	describe('snapshot', () => {
		it('should render the expected html', async () => {
			const page = await newSpecPage({
				components: [OntarioVerticalFormSpacing],
				html: `<ontario-vertical-form-spacing><slot></slot></ontario-vertical-form-spacing>`,
			});
			expect(page.root).toMatchSnapshot();
		});
	});

	it('should render with default spacing if no spacing prop is passed', async () => {
		const page = await newSpecPage({
			components: [OntarioVerticalFormSpacing],
			html: `<ontario-vertical-form-spacing></ontario-vertical-form-spacing>`,
		});

		expect(page.root?.shadowRoot?.querySelector('div')).toHaveClass('ontario-vertical-form-spacing--40');
	});

	it('should render with condensed spacing if the `condensed` spacing prop is passed', async () => {
		const page = await newSpecPage({
			components: [OntarioVerticalFormSpacing],
			html: `<ontario-vertical-form-spacing spacing="condensed"></ontario-vertical-form-spacing>`,
		});

		expect(page.root?.shadowRoot?.querySelector('div')).toHaveClass('ontario-vertical-form-spacing--16');
	});
});
