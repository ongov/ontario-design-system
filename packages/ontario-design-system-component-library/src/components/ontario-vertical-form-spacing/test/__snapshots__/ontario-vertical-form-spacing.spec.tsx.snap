// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ontario-vertical-form-spacing snapshot should render the expected html 1`] = `
<ontario-vertical-form-spacing>
  <template shadowrootmode="open">
    <div class="ontario-vertical-form-spacing--40">
      <slot></slot>
    </div>
  </template>
  <slot></slot>
</ontario-vertical-form-spacing>
`;
